pdh_windows.go

```diff
@@ -0,0 +1,120 @@
+package sigar
+
+/* These are convenience methods for extracting metrics from the Windows
+   Performance Data Helper (PDH) library */
+
+import (
+	"fmt"
+	"github.com/scalingdata/win"
+	"unsafe"
+)
+
+/* Run the set of provided queries once and report back the raw counter values for each.
+   Queries must have a wildcard, like `\Processor(*)\% idle time`. Returns a map of
+   wildcard values to arrays of results. The indices in the array match up with the
+   index of the query in `queries`. */
+func runRawPdhArrayQueries(queries []string) (map[string][]uint64, error) {
+	var query win.PDH_HQUERY
+	counters := make([]win.PDH_HCOUNTER, len(queries))
+	results := make(map[string][]uint64, len(queries))
+	// Open a new query
+	err := win.PdhOpenQuery(0, 0, &query)
+	if err != 0 {
+		return nil, fmt.Errorf("Error opening PDH query: %X", err)
+	}
+
+	// Add all of the counters. If any fail, give up.
+	for i := 0; i < len(queries); i++ {
+		err = win.PdhAddEnglishCounter(query, queries[i], 0, &counters[i])
+		if err != 0 {
+			return nil, fmt.Errorf("Error calling AddCounter for metric %v: %X", queries[i], err)
+		}
+	}
+
+	// Collect query data only once. For raw counters this will give us a value at a point in time
+	err = win.PdhCollectQueryData(query)
+	if err != 0 {
+		return nil, fmt.Errorf("Error collecting data: %X", err)
+	}
+
+	// For each query PDH will hand back an array of key-value pairs
+	for i := 0; i < len(queries); i++ {
+		bufSize := uint32(0)
+		items := uint32(0)
+
+		// Get the size for the raw data buffer
+		err = win.PdhGetRawCounterArray(counters[i], &bufSize, &items, nil)
+		if err != win.PDH_MORE_DATA {
+			win.PdhCloseQuery(query)
+			return nil, fmt.Errorf("Error getting raw array size: %X", err)
+		}
+		buffer := make([]byte, bufSize)
+		err = win.PdhGetRawCounterArray(counters[i], &bufSize, &items, (*win.PDH_RAW_COUNTER_ITEM)(unsafe.Pointer(&buffer[0])))
+		if err != 0 {
+			win.PdhCloseQuery(query)
+			return nil, fmt.Errorf("Error getting raw array: %X", err)
+		}
+
+		// Do some pointer arithmetic to walk through the elements of the array. The whole buffer isn't
+		// PDH_RAW_COUNTER_ITEM structs, there's other data after the structs.
```

Comment:
Does this maybe belong in the `win` library?  Maybe a function that returns the desired mapping?